component/main.jsonnet

```
@@ -1,10 +1,38 @@
 // main template for namespaces
+local com = import 'lib/commodore.libjsonnet';
 local kap = import 'lib/kapitan.libjsonnet';
 local kube = import 'lib/kube.libjsonnet';
 local inv = kap.inventory();
+
 // The hiera parameters for the component
 local params = inv.parameters.namespaces;
+local instanceName = inv.parameters._instance;
+
+// Prevent creating a namespace that is not instantiated
+assert instanceName != 'namespaces' : 'component must be instantiated with a name';
+
+local namespace(name) = {
+  // Prevent creating a namespace that is a reserved Kubernetes namespace
+  assert !std.startsWith(name, 'kube') : 'namespace name must not start with "kube"',
+  assert !std.startsWith(name, 'openshift') : 'namespace name must not start with "openshift"',
+  // Prevent creating a namespace that is a reserved Syn namespace
+  assert !std.startsWith(name, 'syn') : 'namespace name must not start with "syn"',
+
+  apiVersion: 'v1',
+  kind: 'Namespace',
+  metadata: {
+    annotations: {
+      'argocd.argoproj.io/sync-options': 'Delete=false',
+    } + com.makeMergeable(std.get(params.namespaces[name], 'annotations', {})),
+    labels: {
+      name: name,
+    } + com.makeMergeable(std.get(params.namespaces[name], 'labels', {})),
+    name: name,
+  },
+};
 
 // Define outputs below
 {
+  [name]: namespace(name)
+  for name in std.objectFields(params.namespaces)
 }
```

By reading from the merged `parameters.namespaces` inventory parameter, users can still generate conflicts when they define any namespaces as 

```
parameters:
  namespaces:
    namespaces:
      foo: {}
```
since those namespaces will be rendered for each instance. I'd probably do the following here to only render namespaces defined for the current instance:

```suggestion
local instanceKey = std.strReplace(instanceName, '-', '_');
local instanceParams = parameters[instanceKey];
// Define outputs below
{
  [name]: namespace(name)
  for name in std.objectFields(instanceParams.namespaces)
}
```

You can use `instanceParams` in `namespace()` as well, but it's not strictly necessary there.

In theory, you can also add an assertion to prevent users from defining namepaces in the default `namespaces` key with something like:

```
local defaultNamespaces = std.setDiff(std.objectFields(params.namespaces), std.objectFields(instanceParams.namespaces));
assert std.length(defaultNamespaces) == 0 : "Configuring namespaces in `parameters.namespaces.namespaces` isn't allowed, currently: %s" % [ defaultNamespaces ];
```

---

I see